include/swift/SIL/SILBasicBlock.h

```diff
@@ -205,60 +206,56 @@ public llvm::ilist_node<SILBasicBlock>, public SILAllocated<SILBasicBlock> {
-    std::advance(Pos, Index);
-    return insertFunctionArgument(Pos, Ty, OwnershipKind, D);
-  }
-
-  /// Replace the \p{i}th Function arg with a new Function arg with SILType \p
-  /// Ty and ValueDecl \p D.
-  SILFunctionArgument *replaceFunctionArgument(unsigned i, SILType Ty,
-                                               ValueOwnershipKind Kind,
-                                               const ValueDecl *D = nullptr);
-
-  /// Replace the \p{i}th BB arg with a new BBArg with SILType \p Ty and
-  /// ValueDecl \p D.
-  ///
-  /// NOTE: This assumes that the current argument in position \p i has had its
-  /// uses eliminated. To replace/replace all uses with, use
-  /// replacePhiArgumentAndRAUW.
-  SILPhiArgument *replacePhiArgument(unsigned i, SILType type,
-                                     ValueOwnershipKind kind,
-                                     const ValueDecl *decl = nullptr);
-
-  /// Replace phi argument \p i and RAUW all uses.
-  SILPhiArgument *
-  replacePhiArgumentAndReplaceAllUses(unsigned i, SILType type,
-                                      ValueOwnershipKind kind,
-                                      const ValueDecl *decl = nullptr);
-
-  /// Allocate a new argument of type \p Ty and append it to the argument
-  /// list. Optionally you can pass in a value decl parameter.
-  SILPhiArgument *createPhiArgument(SILType Ty, ValueOwnershipKind Kind,
-                                    const ValueDecl *D = nullptr);
-
-  /// Insert a new SILPhiArgument with type \p Ty and \p Decl at position \p
-  /// Pos.
-  SILPhiArgument *insertPhiArgument(arg_iterator Pos, SILType Ty,
-                                    ValueOwnershipKind Kind,
-                                    const ValueDecl *D = nullptr);
-
-  SILPhiArgument *insertPhiArgument(unsigned Index, SILType Ty,
-                                    ValueOwnershipKind Kind,
-                                    const ValueDecl *D = nullptr) {
-    arg_iterator Pos = ArgumentList.begin();
-    std::advance(Pos, Index);
-    return insertPhiArgument(Pos, Ty, Kind, D);
+  SILFunctionArgument *
+  createFunctionArgument(SILType type, const ValueDecl *decl = nullptr,
+                         bool disableEntryBlockVerification = false);
+
+  // Base creation routines for arguments.
+#define ARGUMENT_BOILERPLATE(CLASSNAME, PRETTYNAME)                            \
+  CLASSNAME *replace##PRETTYNAME(unsigned i, SILType type,                     \
+                                 ValueOwnershipKind kind,                      \
+                                 const ValueDecl *decl = nullptr) {            \
+    validateEntryBlock(SILArgumentKind::CLASSNAME);                            \
+    auto *newArg =                                                             \
+        replaceArgument(SILArgumentKind::CLASSNAME, i, type, kind, decl);      \
+    return reinterpret_cast<CLASSNAME *>(newArg);                              \
```

Comment:
The reason I am using reinterpret_cast here instead of cast<...> is that we know what the underlying class is vending meaning the checked cast doesn't really give us anything. Additionally, this makes it so that I can avoid needing the definition of SILArgument in SILBasicBlock.h which would create include circularity issues.